### ISSUE TITLE
Redo Backup v1.0.4 backwards compatibility on AMD64 build

### ISSUE THREAD

**Issue by shasheene** [rescuezilla/rescuezilla#77]
During the development of Rescuezilla v1.0.5 (based on a 32-build Ubuntu 18.04 Bionic), it became clear that changes to 'sfdisk' (partition table backup/restore utility) and 'partclone' made Redo Backup v1.0.4 backwards compatibility difficult to achieve.

For example, using Redo Backup v1.0.4's sfdisk v2.20.1 to backup the partition table of an MBR-formatted disk containing a simple extended partition layout would fail to restore using Redo Backup v1.0.5's sfdisk v2.31.1. The root cause was due to changes in the sfdisk file format. I encountered similar issues with "partclone.restore" detecting the wrong filesystem. I haven't been able to find an example test case that reproduces this partclone.restore issue yet, but use of partclone.restore _is_ discouraged in favor of "partclone.$fs --restore" as "[partclone.restore sometimes might fail to detect the correct file system](https://sourceforge.net/p/partclone/discussion/638474/thread/927b451c73/#77d1)".

For Resceuzilla v1.0.5 the workaround was simply to import the Redo Backup v1.0.4 binaries for sfdisk v2.20.1 and partclone.restore v0.2.43, and use them purely as a means to provide reliable backwards compatibility. See the commit message of 73b5d397916479c934fcf618b5c5cb5e81738606 for full details. This worked well in testing and no issues have yet been reported.

---

For Rescuezilla v1.0.6, the 64-bit version is based on Ubuntu 20.04 Focal, which has dropped 32-bit support. Launching the 32-bit partclone binary produces a cryptic error:

> bash: /usr/sbin/partclone.restore.v0.2.43: No such file or directory

The root cause is that without 32-bit support the i386 partclone binary won't work:

> file /usr/sbin/partclone.restore.v0.2.43
/usr/sbin/partclone.restore.v0.2.43: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux.so.2, for GNU/Linux 2.6.15, BuildID[sha1]=285ed0aea6136e000daef51325099c2698e4596a, stripped

I tried digging up the Ubuntu 12.04 [partclone v0.2.43 AMD64 deb files](http://old-releases.ubuntu.com/ubuntu/pool/universe/p/partclone/) to install on Ubuntu 20.04 but ran into dependency hell. **Given Redo Backup v1.0.4 continues to get ~700 downloads a week, it's vital that the 64-bit version of Rescuezilla v1.0.6 continues to support the ability to restore backups made with Redo Backup v1.0.4.**

It appears the best solution seems to be to build both sfdisk v2.20.1 and partclone v0.2.43 from source, at least for the AMD64 release based on Ubuntu 20.04 Focal. I tried building partclone v0.2.43 from source on a Rescuezilla v1.0.6 development version and while I had to disable a number of build flags (eg, JFS filesystem) the static binary produced worked well.

Ideally, rather than simply dropping in a pre-built binary onto the filesystem again, the Rescuezilla build script will be modified to compile both sfdisk and partclone. This approach has the advantage that future versions of Rescuezilla can more recent versions of partclone rather than being stuck on what Ubuntu provides (important for #65).

However, for the goal of getting Rescuezilla v1.0.6 released as soon as possible, releasing with a pre-built binary is not be a show stopper (but definitely not ideal).

**Comment by zebradots** [comment]
Hi @shasheene,

This issue describes a scenario that to our knowledge has never been encountered when using Redo 2.0 (not Rescuezilla) to restore a backup image created from version 1.0.4 or earlier. In every test we've run, Redo 1.0.x backups are restored flawlessly using Redo 2.0.

While there certainly may have been a situation in which *Rescuezilla* had issues restoring from Redo 1.0.4 backups, we question the characterization found in the footnotes of [this wiki page regarding unofficial Redo versions](https://github.com/rescuezilla/rescuezilla/wiki/Bugs-in-unofficial-Redo-Backup-updates):

> Also Redo Rescue does not yet correctly support restoring the Redo 1.x image format which can only be achieved by providing optional older versions of the partclone and sfdisk utilities for proper backwards compatibility.

This is false. Redo 2.0 is routinely used to restore Redo 1.0.4 backups without issue. We have not received a single report of an issue with this. If there is a specific system configuration or circumstance which reproducibly leads to this problem, please open an issue at https://github.com/redorescue/redorescue/issues.

In any case, it is by no means accurate to suggest that Redo 2.0 is incapable of restoring backup images created with Redo 1.0.x.

**Comment by shasheene** [comment]
Hi,

Please try the reproduction instructions in the first comment of this Github issue (the second paragraph).

I just tried your latest release (Redo Rescue v2.0.3) and confirmed it continues to **not** fully support the _Redo Backup and Recovery_ 1.x image format_ correctly (whilst Rescuezilla supports the format without issue).

I have amended the Wiki page so it reads as "completely correctly" instead of "correctly", which is perhaps a more accurate reflection 

I have actually already informed of this bug 3 times already. Please search for the keyword 'sfdisk' in your [Redo Rescue v2.0.0 announcement thread](https://sourceforge.net/p/redobackup/discussion/general/thread/d0e37c4750). On that forum thread, I mentioned the issue 3 times:

> As mentioned, restoring an image created by v1.0.4 using Redo Rescue v2.0.0 is not a general solution to this relatively common yet high severity "corrupted restore" problem, as many filesystems and partitioning structures will not correctly restore due changes in the internal sfdisk and partclone utilities.

> ...some of which (including v1.0.4) require the bundling of older versions of sfdisk and partclone and carefully managed logic in order to correctly achieve complete backwards compatibility across every possible partitioning structure.

> I also don't want to have to add complexity to Rescuezilla to correctly support the unofficial stop-gap updates and the Redo 1.x image format by eg, [compiling older versions of the partclone and sfdisk utilities for 64-bit systems](https://github.com/rescuezilla/rescuezilla/commit/68aaee87a5780fc1ef6702fe679e30f50b3250df) (#77), but I feel a deep sense of obligation.

To re-iterate, the root cause has nothing to do with Rescuezilla (or _Redo Rescue_ for that matter), instead it's because changes that were made by the authors of the 'sfdisk' (partition table backup/restore utility) and 'partclone' utilities have made to their application in the 7 years since _Redo Backup and Recovery_ was released in November 2012. As mentioned in the first comment of this Github issue:

> For example, using Redo Backup v1.0.4's sfdisk v2.20.1 to backup the partition table of an MBR-formatted disk containing a simple extended partition layout would fail to restore using Redo Backup v1.0.5's sfdisk v2.31.1. The root cause was due to changes in the sfdisk file format. I encountered similar issues with more recent versions of "partclone.restore" detecting the wrong filesystem and therefore failing to restore the filesystem. After a cursory attempt I was not able to find an example test case that reproduces this partclone.restore issue again, but use of partclone.restore is discouraged in favor of "[partclone.$fs --restore" as "partclone.restore sometimes might fail to detect the correct file system](https://sourceforge.net/p/partclone/discussion/638474/thread/927b451c73/#77d1)", which is exactly the issue I experienced during earlier testing.

To re-iterate on what's written above, the backwards compatibility of sfdisk with respect to simple extended partition layouts is not the only problem. Partclone's internal format seems to have changed making partclone.restore far less reliable than in the past. I have definitely manged to reproduce this in an early build of Rescuezilla, but as mentioned in that comment I don't have a readily available test case. I realized this is an issue during the development of Rescuezilla v1.0.5 (my first release) and fixed it by using the old partclone.restore version for backwards compatibility. Since then I haven't seen any problems, it may have been related to a less popular filesystem, I don't recall. I wish I had more information on the partclone issue. It definitely was there though and the most correct solution was using an older partclone.restore binary.

---

So, as part of my Rescuezilla development work, at some point in the (possibly distant) future I intend on systematically looking at _Redo Rescue_ the same way I did with _Redo Backup and Recovery -- including providing thorough user support to any _Redo Rescue_ who wants it (as I already do for Rescuezilla and as I was doing for _Redo Backup and Recovery_ before my systematic user support to the hundreds of open forum topics was all deleted).

I've already spent way too much time trying to explain this issue. Until I reach that point where I start looking at _Redo Rescue_ more deeply, I am focusing my limited time on development and maintenance of Rescuezilla rather than spending my time looking at another project like _Redo Rescue_. To re-iterate, once I start looking at _Redo Rescue_ deeply, I may be opening _Redo Rescue_ Github issues as relevant.

**Comment by shasheene** [comment]
Hi,

Please try the reproduction instructions in the first comment of this Github issue (the second paragraph).

I just tried your latest release (Redo Rescue v2.0.3) and confirmed it continues to **not** fully support the _Redo Backup and Recovery_ 1.x image format correctly (whilst Rescuezilla supports the format without issue).

I have amended the Wiki page so it reads as "completely correctly" instead of "correctly", which is perhaps a more accurate reflection 

I have actually already informed you of this bug 3 times already. Please search for the keyword 'sfdisk' in your [Redo Rescue v2.0.0 announcement thread](https://sourceforge.net/p/redobackup/discussion/general/thread/d0e37c4750). On that forum thread, I mentioned the issue 3 times:

> As mentioned, restoring an image created by v1.0.4 using Redo Rescue v2.0.0 is not a general solution to this relatively common yet high severity "corrupted restore" problem, as many filesystems and partitioning structures will not correctly restore due changes in the internal sfdisk and partclone utilities.

> ...some of which (including v1.0.4) require the bundling of older versions of sfdisk and partclone and carefully managed logic in order to correctly achieve complete backwards compatibility across every possible partitioning structure.

> I also don't want to have to add complexity to Rescuezilla to correctly support the unofficial stop-gap updates and the Redo 1.x image format by eg, [compiling older versions of the partclone and sfdisk utilities for 64-bit systems](https://github.com/rescuezilla/rescuezilla/commit/68aaee87a5780fc1ef6702fe679e30f50b3250df) (#77), but I feel a deep sense of obligation.

To re-iterate, the root cause has nothing to do with Rescuezilla (or _Redo Rescue_ for that matter), instead it's because changes that were made by the authors of the 'sfdisk' (partition table backup/restore utility) and 'partclone' utilities have made to their application in the 7 years since _Redo Backup and Recovery_ was released in November 2012. As mentioned in the first comment of this Github issue: (I have just edited the post to refer to "a prerelease build of Rescuezilla v1.0.5" not the typo "Redo Backup v1.0.5")

> For example, using Redo Backup v1.0.4's sfdisk v2.20.1 to backup the partition table of an MBR-formatted disk containing a simple extended partition layout would fail to restore using a prerelease build of Rescuezilla v1.0.5 with the native sfdisk v2.31.1. The root cause was due to changes in the sfdisk file format. I encountered similar issues with more recent versions of "partclone.restore" detecting the wrong filesystem and therefore failing to restore the filesystem. After a cursory attempt I was not able to find an example test case that reproduces this partclone.restore issue again, but use of partclone.restore is discouraged in favor of "[partclone.$fs --restore" as "partclone.restore sometimes might fail to detect the correct file system](https://sourceforge.net/p/partclone/discussion/638474/thread/927b451c73/#77d1)", which is exactly the issue I experienced during earlier testing.

To re-iterate on what's written above, the backwards compatibility of sfdisk with respect to simple extended partition layouts is not the only problem. Partclone's internal format seems to have changed making partclone.restore far less reliable than in the past. I have definitely manged to reproduce this in an early build of Rescuezilla, but as mentioned in that comment I don't have a readily available test case. I realized this is an issue during the development of Rescuezilla v1.0.5 (my first release) and fixed it by using the old partclone.restore version for backwards compatibility. Since then I haven't seen any problems, it may have been related to a less popular filesystem, I don't recall. I wish I had more information on the partclone issue. It definitely was there though and the most correct solution was using an older partclone.restore binary.

---

So, as part of my Rescuezilla development work, at some point in the (possibly distant) future I intend on systematically looking at _Redo Rescue_ the same way I did with _Redo Backup and Recovery -- including providing thorough user support to any _Redo Rescue_ who wants it (as I already do for Rescuezilla and as I was doing for _Redo Backup and Recovery_ before my systematic user support to the hundreds of open forum topics was all deleted).

I've already spent way too much time trying to explain this issue. Until I reach that point where I start looking at _Redo Rescue_ more deeply, I am focusing my limited time on development and maintenance of Rescuezilla rather than spending my time looking at another project like _Redo Rescue_. To re-iterate, once I start looking at _Redo Rescue_ deeply, I may be opening _Redo Rescue_ Github issues as relevant.

**Comment by zebradots** [comment]
Hi @shasheene,

No worries, and we don't want to waste your time on this, but since it's relevant to both projects, it is worth mentioning that the function `restore_part()` (Redo Rescue 2.0.3) defined at https://github.com/redorescue/redorescue/blob/912772a75ad8118e52fd226b5c50aea9740381fb/overlay/rootdir/var/www/html/functions.inc.php#L784 doesn't call `partclone.restore`, but instead uses `partclone.` followed by the detected filesystem (defaulting to `partclone.dd` if it fails to be detected). In fact, line 795 reads:

`// Use of partclone.restore deprecated; use filesystem-specific binary with "--restore"`

Could the problem lie somewhere else? Because it doesn't appear to call `partclone.restore` anywhere.

**Comment by shasheene** [comment]
It sounds like you are handling the partclone changes very well. I don't have a reproducible test case for any partclone issues, so I'm sure what you've done is fine.

It's trivial to reproduce the sfdisk issue. Create a simple extended partition table structure on _Redo Backup and Recovery_ v1.0.4 then restore it using _Redo Rescue_ v2.0.3. Te partition table doesn't get restored correctly, due to changes in the sfdisk format.

**Comment by zebradots** [comment]
Hi @shasheene! First, thank you for taking the time to explain this in such great detail. We appreciate your concern and would certainly like to ensure that any images created with our Redo 1.0.x releases (with the exception of the unofficial user-contributed ISOs) can safely be restored using our latest version (2.0.3 as of this writing). While most of the info below refers to Redo 2.0.x, this core functionality is central to Rescuezilla as well, and is critical for a reliable recovery.

As explained in the previous post, we haven't yet encountered a production or test case that has resulted in an error simply because it contains an extended partition. But just to be sure, we completed some more tests, including the one described here. First, we booted Redo 1.0.4 and created a typical table containing four data partitions as shown:

1. sda1: ext3 (primary)
2. sda2: (extended)
3. sda5: ext4 (logical)
4. sda6: fat32 (logical)
5. sda7: ntfs (logical)

Each partition was then formatted with its respective filesystem and verified. The Redo 1.0.4 application was then used to create a backup image of the drive to a local NAS. After this, the drive was erased with `dd if=/dev/zero of=/dev/sda bs=1M count=100; sync` and rebooted into Redo 2.0.3 to confirm that no partition table existed. The backup image was then restored from the image stored on the network shared drive with Redo 2.0.3, which reported success without any error messages. We confirmed all four filesystems were valid and intact by mounting and viewing the directory listings for each, and all is well. The restore works as expected.

For reference, the original `.sfdisk` file from the 1.0.4 backup reads:

```
# partition table of /dev/sda
unit: sectors

/dev/sda1 : start=     2048, size= 10240000, Id=83
/dev/sda2 : start= 10242048, size= 31700992, Id= 5
/dev/sda3 : start=        0, size=        0, Id= 0
/dev/sda4 : start=        0, size=        0, Id= 0

/dev/sda5 : start= 10244096, size= 10240000, Id=83
    -     : start= 20484096, size= 10242048, Id= 5
    -     : start= 10242048, size=        0, Id= 0
    -     : start= 10242048, size=        0, Id= 0

/dev/sda6 : start= 20486144, size= 10240000, Id= b
    -     : start= 30726144, size= 11216896, Id= 5
    -     : start= 20484096, size=        0, Id= 0
    -     : start= 20484096, size=        0, Id= 0

/dev/sda7 : start= 30728192, size= 11214848, Id= 7
    -     : start= 30726144, size=        0, Id= 0
    -     : start= 30726144, size=        0, Id= 0
    -     : start= 30726144, size=        0, Id= 0
```

So again, we greatly appreciate your input and definitely want to ensure that Redo 2.0.x restores legacy backup images, so we've been paying close attention to your reports detailed above. Yet we just aren't able to reproduce the issue using the steps described above. Not to say there isn't _some_ scenario that may trigger an error, but given the results, it doesn't appear trivial to reproduce.

This, coupled with many successful restores in production that make use of similar partition layouts, is why we objected to the notion that Redo 1.0.x backups simply couldn't be restored with 2.0.x releases. We're genuinely unable to reproduce this error, despite a battery of tests covering many different configurations and versions. This isn't to say there are no bugs, but if there _is_ an issue here, more details are required to reproduce it. It could stem from something other than changes to `sfdisk` and `partclone`, (which one would assume are reasonably backward compatible), and may be related to the integrity of the filesystem itself, if we had to venture a guess. If it turns out to be unrelated to `sfdisk` or `partclone` changes, it would be a welcome discovery, as it alleviates the need to compile and package the legacy versions in Rescuezilla.

Your insight is appreciated, as always, and thanks again for taking the time to explain your experience!

**Comment by shasheene** [comment]
The issue is you have only erased the Master Boot Record. On traditional DOS (MBR) formatted disks the extended partitioning information resides in the [Extended Boot Record](https://en.wikipedia.org/wiki/Extended_boot_record), which in your case is located within the /dev/sda2 logical partition. In your workflow, after erasing the MBR you can erase the EBR by running `dd if=/dev/zero of=/dev/sda2 bs=1M count=100; sync`.

I haven't looked into the _Redo Rescue_ log files, but based on my aforementioned previous analysis of the bugs in [louvetch's unofficial Ubuntu 16.04 _Redo Backup and Recovery_ update](https://github.com/rescuezilla/rescuezilla/wiki/Bugs-in-unofficial-Redo-Backup-updates#bugs-in-louvetchs-ubuntu-1604-releases), I can tell you with 100% certainty that _Redo Rescue_'s is suffering from a similar bug in that your restore `sfdisk` invocation on a restore is not actually having any affect at all.

You're restoring your partitioning information from the MBR backup, but not restoring any extended partitioning information. As mentioned, the _only_ solution I am aware of to this problem is to provide an older sfdisk binary for backwards compatibility.

**Comment by zebradots** [comment]
Your last comment was exactly correct, of course! When the extended partition details are wiped from `/dev/sda2`, the error is 100% reproducible. Thank you for pointing us to the exact cause, @shasheene. We're sorry for the headaches this probably caused, though as you pointed out, it's really the result of changes to `sfdisk` over the years.

Luckily, the difference between the legacy `sfdisk` output and the output from modern binaries appears to be simply a matter of formatting. Here's a one-liner we came up with to fix the formatting of a file dumped by legacy `sfdisk` versions:

`perl -p -e 's/^    \-.*$//' backup_name.sfdisk | grep "\S"`

Running this command on the `.sfdisk` file of a Redo 1.0.4 backup will remove the extraneous lines that seem to confuse newer versions of `sfdisk` as to where some partitions are located. After we ran this command to clean the old file up, we directed straight back into the command with `sfdisk --force /dev/sda < new.sfdisk; partprobe` and confirmed that all the partitions were picked up in Gparted. It seems to work fine. The regex shouldn't break modern `sfdisk` outputs if it's applied to them, but at least in Redo, logic for handling legacy backups resides in a separate function so it can selectively apply this additional step to prepare the file before restoring from an old backup.

Thanks again! Hopefully this info may alleviate the need to compile and maintain old versions of `sfdisk` for Rescuezilla, although compiling the old version is certainly prudent and safe.
